delta/plugins/project-deletion/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/projectdeletion/ProjectDeletion.scala

```diff
@@ -0,0 +1,133 @@
+import ch.epfl.bluebrain.nexus.delta.sdk.model.search.SearchParams.ProjectSearchParams
+import ch.epfl.bluebrain.nexus.delta.sdk.plugin.Plugin
+import ch.epfl.bluebrain.nexus.delta.sdk.{ProjectReferenceFinder, ProjectResource, Projects}
+import ch.epfl.bluebrain.nexus.delta.sourcing.EventLog
+import ch.epfl.bluebrain.nexus.delta.sourcing.config.DatabaseFlavour
+import com.datastax.oss.driver.api.core.uuid.Uuids
+import com.typesafe.scalalogging.Logger
+import fs2.Stream
+import monix.bio.{Fiber, IO, Task, UIO}
+
+import java.time.Instant
+import scala.concurrent.duration.DurationInt
+
+class ProjectDeletion(fiber: Fiber[Throwable, Unit]) extends Plugin {
+
+  private val logger: Logger = Logger[ProjectDeletion]
+
+  /**
+    * Stop the plugin. This should allow the plugin to terminate gracefully.
+    */
+  override def stop(): Task[Unit] = fiber.cancel.timeout(10.seconds).flatMap {
+    case Some(()) => UIO.delay(logger.info("Project Deletion plugin stopped."))
+    case None     => UIO.delay(logger.error("Project Deletion plugin could not be stopped in 10 seconds."))
+  }
+}
+
+object ProjectDeletion {
+
+  def apply(
+      projects: Projects,
+      config: ProjectDeletionConfig,
+      serviceAccount: ServiceAccount,
+      eventLog: EventLog[Envelope[ProjectScopedEvent]]
+  )(implicit prf: ProjectReferenceFinder): Task[ProjectDeletion] = {
+
+    implicit val caller: Caller = serviceAccount.caller
+    val logger: Logger          = Logger[ProjectDeletion]
+
+    // a filter on projects that depends on the deprecation configuration
+    def deprecationFilter(pr: ProjectResource): Boolean =
+      if (config.deleteDeprecatedProjects) pr.deprecated
+      else true
+
+    def lastUpdatedFilter(pr: ProjectResource, now: Instant): Boolean =
+      (now.getEpochSecond - pr.updatedAt.getEpochSecond) > config.idleInterval.toSeconds
+
+    def offset(now: Instant): Offset =
+      eventLog.flavour match {
+        case DatabaseFlavour.Postgres  => NoOffset
+        case DatabaseFlavour.Cassandra => TimeBasedUUID(Uuids.startOf(now.toEpochMilli - config.idleInterval.toMillis))
+      }
+
+    def deleteProject(pr: ProjectResource): UIO[Unit] =
+      projects
+        .delete(pr.value.ref, pr.rev)
+        .flatMap { case (uuid, _) =>
+          UIO.delay(logger.info(s"Marked project '${pr.value.ref}' for deletion ($uuid)."))
+        }
+        .onErrorHandleWith { rejection =>
+          UIO.delay(
+            logger.error(
+              s"Unable to mark project '${pr.value.ref}' for deletion due to '${rejection.reason}'. It will be retried at the next pass."
+            )
+          )
+        }
+
+    def checkAndDeleteProject(pr: ProjectResource, now: Instant): Task[Unit] = {
+      implicit val rm: Mapper[ProjectNotFound, Throwable] =
+        Mapper(_ => new IllegalArgumentException(s"Project '${pr.value.ref}' was not found."))
+      EventLogUtils
+        .currentProjectEvents(projects, eventLog, pr.value.ref, offset(now))
```

Comment:
Do we need to do it this way? Can't we rather make calls to the projects statistics (I believe this is now a call to elasticsearch instead of a cache, but it still going to be more efficient than get all the events)

Reply:
We discussed this with bogdan and it is safer to rely on the primary store to do this task as for some reason stats may be wrong (replay, ...)